### PR TITLE
feat: allowlist host param using msg typeURLs

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding allowlist gov param for limiting host chain ICA functionality.

closes: #511 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/ibc/applications/interchain_accounts/host/v1/host.proto [hunk 1]
```diff
@@ -10,6 +10,8 @@ import "gogoproto/gogo.proto";
 // Params defines the set of on-chain interchain accounts parameters.
 // The following parameters may be used to disable the host submodule.
 message Params {
-  // host_enabled enables or disables the host submodule. 
+  // host_enabled enables or disables the host submodule.
   bool host_enabled = 1 [(gogoproto.moretags) = "yaml:\"host_enabled\""];
+  // allow_messages defines a list of sdk message typeURLs allowed to be executed on a host chain.
+  repeated string allow_messages = 2 [(gogoproto.moretags) = "yaml:\"allow_messages\""];
```
Comment (damiannolan) [inline proto/ibc/applications/interchain_accounts/host/v1/host.proto:16]: I named this `allow_messages` as I thought `allow_list` was maybe a little ambiguous. 
Open to renaming

File: proto/ibc/applications/interchain_accounts/v1/genesis.proto [hunk 2]
```diff
@@ -22,7 +22,7 @@ message ControllerGenesisState {
       [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"active_channels\""];
   repeated RegisteredInterchainAccount interchain_accounts = 2
       [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"interchain_accounts\""];
-  repeated string ports = 3;
+  repeated string                                           ports  = 3;
```
Comment (damiannolan) [inline proto/ibc/applications/interchain_accounts/v1/genesis.proto:25]: `make proto-all` decided to reformat this

File: modules/apps/27-interchain-accounts/host/types/params.go [hunk 3]
```diff
@@ -57,3 +65,12 @@ func validateEnabled(i interface{}) error {
 
 	return nil
 }
+
+func validateAllowlist(i interface{}) error {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/types/params.go:69]: Is it possible to have a []string slice with a empty entry?

Such as: []string{bankURL, "", stakingURL}?

File: modules/apps/27-interchain-accounts/host/types/keys.go [hunk 4]
```diff
@@ -1,9 +1,24 @@
...
 const (
 	// ModuleName defines the interchain accounts host module name
 	ModuleName = "icahost"
 
 	// StoreKey is the store key string for the interchain accounts host module
 	StoreKey = ModuleName
 )
+
+// ContainsType returns true if the sdk.Msg TypeURL is present in allowMsgs, otherwise false
+func ContainsType(allowMsgs []string, msg sdk.Msg) bool {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/types/keys.go:16]: `ContainsMsgType`?

File: modules/apps/27-interchain-accounts/host/ibc_module_test.go [hunk 5]
```diff
@@ -436,6 +437,9 @@ func (suite *InterchainAccountsTestSuite) TestOnRecvPacket() {
 			}
 			packetData = icaPacketData.GetBytes()
 
+			expParams := hosttypes.NewParams(true, []string{sdk.MsgTypeURL(msg)})
+			suite.chainB.GetSimApp().ICAHostKeeper.SetParams(suite.chainB.GetContext(), expParams)
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/ibc_module_test.go:441]:
```suggestion
			params := hosttypes.NewParams(true, []string{sdk.MsgTypeURL(msg)})
			suite.chainB.GetSimApp().ICAHostKeeper.SetParams(suite.chainB.GetContext(), params)
```

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go [hunk 6]
```diff
@@ -49,6 +50,9 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 				}
 
 				packetData = icaPacketData.GetBytes()
+
+				expParams := types.NewParams(true, []string{sdk.MsgTypeURL(msg)})
+				suite.chainB.GetSimApp().ICAHostKeeper.SetParams(suite.chainB.GetContext(), expParams)
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/keeper/relay_test.go:55]: ditto on renaming `expParams` to `params`?

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go [hunk 7]
```diff
@@ -309,6 +334,27 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 			},
 			false,
 		},
+		{
+			"unauthorised: message type not allowed",
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/keeper/relay_test.go:338]: maybe add a comment that we explicitly don't update the host params

**Review by colin-axner — APPROVED** [review]
ACK pending passing build/tests

Great work! Left some suggestions, but no requested changes

File: modules/apps/27-interchain-accounts/host/types/params.go (hunk 3 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/host/types/params.go:69]: Good point, added a check

File: modules/apps/27-interchain-accounts/host/types/keys.go (hunk 4 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/host/types/keys.go]: Done!

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go (hunk 6 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go]: Resolved all of these, copy pasta from another file :))

File: modules/apps/27-interchain-accounts/host/keeper/relay_test.go (hunk 7 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/host/keeper/relay_test.go]: Added!

File: modules/apps/27-interchain-accounts/host/types/params.go [hunk 8]
```diff
@@ -57,3 +65,18 @@ func validateEnabled(i interface{}) error {
...
+	allowMsgs, ok := i.([]string)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", i)
+	}
+
+	for _, typeURL := range allowMsgs {
+		if typeURL == "" {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/types/params.go:76]:
```suggestion
		if strings.TrimSpace(typeURL) == "" {
```
you'll need to add strings import as well

Comment (damiannolan) [reply modules/apps/27-interchain-accounts/host/types/params.go:76]: Ah good spot! Lemme fix this :)

**Comment by seantking** [comment]
Nice :handshake: 

I guess we should discuss what the default params for the `allow_messages` are?

**Review by seantking — APPROVED** [review]

**Pull request merged**